### PR TITLE
Gci 1329 handle mongo order saving failure with http 500

### PR DESCRIPTION
**Pull request by rjadhav-ch**
This adds exception handling to handle `MongoException` when order saving fails due to a Mongo failure. This exception is wrapped in `MongoOperationException` to allow selective application when handled via `ExceptionHandler`.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java [hunk 1]
```diff
@@ -67,7 +69,16 @@ public Order createOrder(final Checkout checkout) {
         LOGGER.info("Publishing notification to Kafka 'order-received' topic for order - " + mappedOrder.getId(), logMap);
         sendOrderReceivedMessage(mappedOrder.getId());
 
-        return repository.save(mappedOrder);
+        Order savedOrder = null;
+        try {
+            savedOrder = repository.save(mappedOrder);
+        } catch (MongoException ex) {
+            String errorMessage = String.format("Failed to save order with id %s", mappedOrder.getId());
+            LOGGER.error(errorMessage, ex);
```
Comment (hwilliams1) [inline src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:77]: Couldn't we log this and include the logMap containing the value? e.g.
`LOGGER.error(errorMessage, ex, logMap)`

File: src/test/java/uk/gov/companieshouse/orders/api/service/OrderServiceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import uk.gov.companieshouse.orders.api.model.Order;
+import uk.gov.companieshouse.orders.api.model.OrderData;
+import uk.gov.companieshouse.orders.api.repository.OrderRepository;
+
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@SpringBootTest
+@EmbeddedKafka
+public class OrderServiceIntegrationTest {
```
Comment (hwilliams1) [inline src/test/java/uk/gov/companieshouse/orders/api/service/OrderServiceIntegrationTest.java:26]: Is this a correct name for this test? The test below appears to just be a unit test using mockito?

Comment (rjadhav-ch) [reply src/test/java/uk/gov/companieshouse/orders/api/service/OrderServiceIntegrationTest.java:26]: Good point.

File: src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java [hunk 3]
```diff
@@ -67,7 +69,16 @@ public Order createOrder(final Checkout checkout) {
         LOGGER.info("Publishing notification to Kafka 'order-received' topic for order - " + mappedOrder.getId(), logMap);
         sendOrderReceivedMessage(mappedOrder.getId());
 
-        return repository.save(mappedOrder);
+        Order savedOrder = null;
+        try {
+            savedOrder = repository.save(mappedOrder);
+        } catch (MongoException ex) {
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: I see `MongoException` is an unchecked exception too. What is your thinking behind introducing a new unchecked exception class `MongoOperationException`?

Comment (rjadhav-ch) [reply src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: Wrapping `MongoException` in `MongoOperationException` allows for the exception checking to be selectively applied to specific blocks of code and handled via Spring's `ExceptionHandler` mechanism.

Comment (VaughanJackson123) [reply src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: What `MongoException` scenarios would we like to exclude then? Spring's `ExceptionHandler` mechanism can be used for `MongoException` too?

Comment (rjadhav-ch) [reply src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: `MongoException` is the direct parent of the following exception classes - `MongoChangeStreamException`, `MongoClientException`, `MongoExecutionTimeoutException`, `MongoGridFSException`, `MongoIncompatibleDriverException`, `MongoInternalException`, `MongoInterruptedException`, `MongoServerException`, `MongoSocketException`, so having it directly handled via `ExceptionHandler` would mean returning a 500 status for db query operations that timeout and any other operation that is likely to throw any of these exceptions.

Comment (VaughanJackson123) [reply src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: OK, I was thinking that any of these errors would likely mean that the handling of the current request would invalidate its handling therefore a 500 would be appropriate anyway. Do we now what will happen at the moment if such a DB read error should occur?
I do see that in handling this exception that we now is originated at the point of saving the order we are identifying the exact scenario we are dealing with. 👍

Comment (rjadhav-ch) [reply src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: Without `MongoException` being handled (via `ExceptionHandler`) for a read operation, any failure will likely be handled by checks in the the respective controller method and be dealt with using a 40x error.

Comment (VaughanJackson123) [reply src/main/java/uk/gov/companieshouse/orders/api/service/OrderService.java:75]: Ah maybe not ideal then if it suggests there is something wrong with the request when there is a system issue.

**Pull request merged**